### ISSUE TITLE
Plain English Protocol Documentation

### ISSUE THREAD

**Issue by nbering** [hashicorp/go-plugin#27]
I might be missing some detail, but I'm having a really difficult time following the code to try and understand the protocol used here. I'm not a Go-lang developer, and I'm trying to implement the protocol used in node for that very reason.

I've gotten as far as the initial handshake and then the code just loses all meaning to me. Too many levels of indirection to other libraries. It might have more meaning to someone more familiar with the Go stack.

Here's what I've been able to follow so far:

- Host spawns/forks process for plugin.
- Plugin checks environment variables for "Magic Cookie" to ensure it's hosted by the parent application.
- Plugin listens for unix/tcp socket in somewhat random location/port.
- Plugin reports over stdout to host with message formatted like:
  - `BaseProtocolVersion|AppPluginProtocolVersion|SocketProtocol|SocketAddress`

After that I'm lost in multiplexers, redirected stdio streams, and go-specific network implementations. Once the plugin process starts listening on tcp/unix domain socket, I presume that the host application makes a connection to the socket. Who talks next, and what is the protocol?

Is there a flowchart block diagram somewhere? Anything that describes the control flow for this protocol?

I don't expect a 100-page manual, but some next steps would be helpful. I'd be happy to produce some documentation for other people making this attempt if someone can help me get over this part of the learning curve.

**Comment by nbering** [comment]
I've managed to trace some of this down to the core RPC implementation, which uses the go-lang [net/rpc](https://godoc.org/net/rpc) package. Internally, this uses the [gob](https://godoc.org/encoding/gob) encoding format for transport. There's an HTTP implementation for `net/rpc` as well, but I don't see any evidence that this is being used.

From https://www.terraform.io/docs/internals/internal-plugins.html#supporting-our-community

>In theory, because the plugin interface is HTTP, you could even develop a plugin using a completely different programming language! (Disclaimer, you would also have to re-implement the plugin API which is not a trivial amount of work.)

I'm willing to put in some work to get the HTTP interface working, if there is indeed one. However, if the RPC layer is using the gob encoding as I suspect it is, then this is completely impractical to implement from another language. The protocol is very low-level and depends a lot on the go-lang type system and reflection. To build this in node, I would also need to build out the necessary parser infrastructure to handle Go's native data structures. It's not impossible, but I can't seem to find any work to build on to implement a gob-based RPC library in another language. There is a description of the gob format on GoDoc, but it's pretty hard to follow if you're not used to building low-level protocols.

I guess my next step is to confirm whether the transport protocol is HTTP, gob, or something else.

**Comment by mitchellh** [comment]
Yes I think a plain english guide would be useful. 

I've actually been thinking about what a non-Go plugin would look like recently. I think its possible but at the moment very not straightforward. That's something I want to continue to explore.

Beyond that, I don't have time to fill in a response at the moment but I'll try to get to writing a guide. :)

**Comment by nbering** [comment]
It doesn't look like it's possible at the moment. I looked a little deeper at writing a gob-data decoder, but it's pretty tightly bound to go-lang's type system. Would be seriously challenging to write in any other language.

**Comment by mitchellh** [comment]
Write, that's not how I'd go about doing it. I'd use something more standard like protobufs (as an example). The bigger issue actually is going to be the multiplexer on top of it which is using yamux. While that is tried and true in Go, I think it would be way, way too much to ask to reimplement that in other languages.

That's what I meant by "not very straightforward": I think what I'd rather do is actually change the core of go-plugin to basically have two modes: Go mode (gob + yamux) and non-Go mode (something else). More research necessary, but I think its possible with minimal API changes on the go-plugin front.

**Comment by nbering** [comment]
That's interesting to note. Yamux was part of my confusion in following it all. I can't find an exaplaination of where it actually fits into the larger picture. I know what multi-plexing is, but almost every layer (Terraform, go-plugin, yamux) has something that's doing multiplexing.

**Comment by mitchellh** [comment]
All multiplexing leads down to Yamux. :) 

Yamux allows us to only launch one process per plugin which then only requires a single network connection. All other plugin connections are multiplexed (via Yamux) onto that connection. Further, this is how returning interfaces or streams is enabled. A function that returns an interface like `io.Writer` is actually allocating a Yamux stream, and a new RPC connection is created over that. The thing brokering all this access is the "MuxBroker" that you've probably seen thrown around.

Overall, its actually not used super often. We try to design our own APIs that use go-plugin to avoid returning or passing interfaces if possible since it complicates the understanding significantly.

**Comment by nbering** [comment]
Understood. Thanks for the info, that helps a lot. I think I'm out of time for this weekend, but there's a node-module in my future if I can figure out how to use this without gob. I might be able to do it as a proof-of-concept just by modifying go-plugin to use proto-buf or jsonrpc.

**Comment by nbering** [comment]
Still working on a NodeJS Yamux port so that I can later build a go-plugin NodeJS port. I spent most of last weekend on it. I've got it in a private repo until I figure out what license to put it under.

Currently my version of Yamux accepts connections, parses frame headers, and chunks data from the frame to push to a stream.

Still working out the details of the Yamux protocol. Specifically, the use of window-update frames is a little confusing.

I'm also working out some of the kinks converting concepts like goroutines and channels into NodeJS single-threaded event-loop and callback paradigms. It actually seems like it's going to make for a nice API in the end, but not everything maps one-to-one so reverse-engineering the parts of the protocol implementation that are not in the spec is a little challenging (in a fun way).

**Comment by mitchellh** [comment]
@nbering I added support for gRPC plugins to go-plugin: https://github.com/hashicorp/go-plugin/pull/29

This isn't a drop-in feature. The projects (such as Terraform) will have to explicitly enable support for it. But hopefully we can do that and then you can very easily write your plugins in any language that supports gRPC, which is effectively every major language.

**Comment by nbering** [comment]
@mitchellh Awesome! Thanks! I'll do some more tinkering this weekend.

Good work to all the team at Hashicorp for the amazing work going on with Terraform right now. I'm really impressed with the pace of development.

**Comment by tisba** [comment]
👋

I stumbled across this issue as I was researching how to integrate a Go-lang host with a Node.js guest/plugin. #29 seems to be quite helpful, but I was wondering what happened to your attempted of writing a Node-based plugin, @nbering?

**Comment by nbering** [comment]
@tisba The main hangup was https://github.com/hashicorp/yamux/

Yamux sits as a protocol layer between the plugin implementation, and the socket that Terraform communicates with. But it basically implements TCP on top of TCP - and writing network protocol code was too much for a weekend project. 😞

In the end, I wrote my provider in go-lang as the path of least resistance. I'd love to go back and visit this, but the time cost is a bit high since I've already accomplished my original goal of writing a provider.

**Comment by jefferai** [comment]
I assume using the gRPC method instead would be easier...might want to check that out.

**Comment by nbering** [comment]
@jefferai It's my understanding that Yamux is still being used with gRPC. What's being replaced is the [gob](https://golang.org/pkg/encoding/gob/) serialization protocol that is used for the actual message formatting. That had issues because no language other than go-lang has a gobs implementation, and it's built on the go-lang type system so you'd have to do a lot of plumbing to make it work anywhere else.

In either case, Yamux is still used to multiplex the connection, regardless of the message format.

**Comment by dexterlakin** [comment]
did anyone here ever get to a working example of a plugin implemented in node? any chance I can have a look?
